{"Id": "66387610", "PostTypeId": "1", "AcceptedAnswerId": "66387833", "CreationDate": "2021-02-26T14:27:48.217", "Score": "3", "ViewCount": "2679", "Body": "<p>I have a DynamoDB table created with this Terraform:</p>\n<pre><code>resource &quot;aws_dynamodb_table&quot; &quot;materials_table&quot; {\n  name              = &quot;materials&quot;\n  hash_key          = &quot;MATERIAL&quot;\n  billing_mode      = &quot;PROVISIONED&quot;\n  read_capacity     = 5\n  write_capacity    = 5\n  \n  attribute {\n    name = &quot;MATERIAL&quot;\n    type = &quot;S&quot;\n  }\n}\n</code></pre>\n<p>The table was successfully populated (with 4 records, as noted in <a href=\"https://stackoverflow.com/questions/66382143/lambda-only-inserts-4-records-but-iterates-248-times/66382584#66382584\">this post</a>) but in order to solve the problem (in that post) I have added a field <code>PK</code> and set that as the <code>hash_key</code> field, with this:</p>\n<pre><code>resource &quot;aws_dynamodb_table&quot; &quot;materials_table&quot; {\n  name              = &quot;materials&quot;\n  hash_key          = &quot;PK&quot;\n  billing_mode      = &quot;PROVISIONED&quot;\n  read_capacity     = 5\n  write_capacity    = 5\n  \n  attribute {\n    name = &quot;PK&quot;\n    type = &quot;S&quot;\n  }\n}\n</code></pre>\n<p>This has caused the following error, when running <code>terraform apply</code>:</p>\n<pre><code>Error: error creating DynamoDB Table: ResourceInUseException: Table already exists: materials\n</code></pre>\n<p>What do I need to do in the <code>.tf</code> to get the change accepted?</p>\n", "OwnerUserId": "71376", "LastActivityDate": "2021-02-26T14:58:25.823", "Title": "Changing hash_key with Terraform causes Table already exists error", "Tags": "<amazon-web-services><amazon-dynamodb><terraform-provider-aws><terraform0.14.7>", "AnswerCount": "1", "CommentCount": "0", "ContentLicense": "CC BY-SA 4.0", "links": [{"Id": "1813312651", "CreationDate": "2021-02-26T14:27:48.217", "PostId": "66387610", "RelatedPostId": "66382143", "LinkTypeId": "1"}], "history": [{"Id": "241620829", "PostHistoryTypeId": "2", "PostId": "66387610", "RevisionGUID": "a1da0513-ab4c-4f96-bfee-af0d819d39cb", "CreationDate": "2021-02-26T14:27:48.217", "UserId": "71376", "Text": "I have a DynamoDB table created with this Terraform:\r\n\r\n    resource \"aws_dynamodb_table\" \"materials_table\" {\r\n      name              = \"materials\"\r\n      hash_key          = \"MATERIAL\"\r\n      billing_mode      = \"PROVISIONED\"\r\n      read_capacity     = 5\r\n      write_capacity    = 5\r\n      \r\n      attribute {\r\n        name = \"MATERIAL\"\r\n        type = \"S\"\r\n      }\r\n    }\r\n\r\nThe table was successfully populated (with 4 records, as noted in [this post][1]) but in order to solve the problem (in that post) I have added a field `PK` and set that as the `hash_key` field, with this:\r\n\r\n    resource \"aws_dynamodb_table\" \"materials_table\" {\r\n      name              = \"materials\"\r\n      hash_key          = \"PK\"\r\n      billing_mode      = \"PROVISIONED\"\r\n      read_capacity     = 5\r\n      write_capacity    = 5\r\n      \r\n      attribute {\r\n        name = \"PK\"\r\n        type = \"S\"\r\n      }\r\n    }\r\n\r\n This has caused the following error, when running `terraform apply`:\r\n\r\n    Error: error creating DynamoDB Table: ResourceInUseException: Table already exists: materials\r\n\r\nWhat do I need to do in the `.tf` to get the change accepted?\r\n\r\n\r\n\r\n  [1]: https://stackoverflow.com/questions/66382143/lambda-only-inserts-4-records-but-iterates-248-times/66382584#66382584", "ContentLicense": "CC BY-SA 4.0", "filtered-sentences": [{"source": "Text", "text": "I have a DynamoDB table created with this Terraform: resource \"aws_dynamodb_table\" \"materials_table\" { name = \"materials\" hash_key = \"MATERIAL\" billing_mode = \"PROVISIONED\" read_capacity = 5 write_capacity = 5 attribute { name = \"MATERIAL\" type = \"S\" } } ", "keywords": ["bill"]}, {"source": "Text", "text": "I have added a field `PK` and set that as the `hash_key` field, with this: resource \"aws_dynamodb_table\" \"materials_table\" { name = \"materials\" hash_key = \"PK\" billing_mode = \"PROVISIONED\" read_capacity = 5 write_capacity = 5 attribute { name = \"PK\" type = \"S\" } } ", "keywords": ["bill"]}, {"source": "Text", "text": "This has caused the following error, when running `terraform apply`: Error: error creating DynamoDB Table: ResourceInUseException: Table already exists: materials What do I need to do in the `.tf` to get the change accepted? [1]: https://stackoverflow.com/questions/66382143/lambda-only-inserts-4-records-but-iterates-248-times/66382584#66382584", "keywords": ["change"]}]}, {"Id": "241620830", "PostHistoryTypeId": "1", "PostId": "66387610", "RevisionGUID": "a1da0513-ab4c-4f96-bfee-af0d819d39cb", "CreationDate": "2021-02-26T14:27:48.217", "UserId": "71376", "Text": "Changing hash_key with Terraform causes Table already exists error", "ContentLicense": "CC BY-SA 4.0", "filtered-sentences": [{"source": "Text", "text": "Changing hash_key with Terraform causes ", "keywords": ["change"]}]}, {"Id": "241620831", "PostHistoryTypeId": "3", "PostId": "66387610", "RevisionGUID": "a1da0513-ab4c-4f96-bfee-af0d819d39cb", "CreationDate": "2021-02-26T14:27:48.217", "UserId": "71376", "Text": "<amazon-web-services><amazon-dynamodb><terraform-provider-aws><terraform0.14.7>", "ContentLicense": "CC BY-SA 4.0", "filtered-sentences": []}], "answers": [{"Id": "66387833", "PostTypeId": "2", "ParentId": "66387610", "CreationDate": "2021-02-26T14:43:28.863", "Score": "3", "Body": "<p>Changing some of the attributes in DynamoDB is not permitted, for example, changing Partition key, adding a Local Secondary Index, etc.</p>\n<p>When such changes occur, it will need to replace the resource and to replace, it will try to delete and re-create the resource. During this process, if table already exists, it will fail.</p>\n<p>Only option is to delete the stack or manually delete the DynamoDB Table and let template create it again. Or Renaming the table.</p>\n<p><a href=\"https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/dynamodb_table\" rel=\"nofollow noreferrer\">Documentation</a> says it will force new resource</p>\n<blockquote>\n<p>hash_key - (Required, Forces new resource) The attribute to use as the\nhash (partition) key.</p>\n</blockquote>\n", "OwnerUserId": "10186808", "LastEditorUserId": "10186808", "LastEditDate": "2021-02-26T14:58:25.823", "LastActivityDate": "2021-02-26T14:58:25.823", "CommentCount": "6", "ContentLicense": "CC BY-SA 4.0", "comments": [{"Id": "117366640", "PostId": "66387833", "Score": "0", "Text": "I'm using Terraform, not CloudFormation.", "CreationDate": "2021-02-26T14:45:34.587", "UserId": "71376", "ContentLicense": "CC BY-SA 4.0", "filtered-sentences": []}, {"Id": "117366676", "PostId": "66387833", "Score": "0", "Text": "I was expecting the table to be deleted, but the error implies this is not being attempted.", "CreationDate": "2021-02-26T14:46:20.940", "UserId": "71376", "ContentLicense": "CC BY-SA 4.0", "filtered-sentences": []}, {"Id": "117366711", "PostId": "66387833", "Score": "0", "Text": "I think, it will first try to create the table and then delete it.", "CreationDate": "2021-02-26T14:47:32.520", "UserId": "10186808", "ContentLicense": "CC BY-SA 4.0", "filtered-sentences": []}, {"Id": "117366783", "PostId": "66387833", "Score": "0", "Text": "So renaming the table should help delete it..?", "CreationDate": "2021-02-26T14:50:16.793", "UserId": "71376", "ContentLicense": "CC BY-SA 4.0", "filtered-sentences": []}, {"Id": "117366802", "PostId": "66387833", "Score": "1", "Text": "100% . renaming will surely help.", "CreationDate": "2021-02-26T14:51:13.630", "UserId": "10186808", "ContentLicense": "CC BY-SA 4.0", "filtered-sentences": []}, {"Id": "117367020", "PostId": "66387833", "Score": "0", "Text": "I saw that the docs say \"forces new resource\" and thought this would do exactly that. I guess it's the process it uses to get there which failed.", "CreationDate": "2021-02-26T15:00:03.660", "UserId": "71376", "ContentLicense": "CC BY-SA 4.0", "filtered-sentences": []}], "history": [{"Id": "241621732", "PostHistoryTypeId": "2", "PostId": "66387833", "RevisionGUID": "c74f68c7-aab0-451c-9ea2-b44c1dcd5e61", "CreationDate": "2021-02-26T14:43:28.863", "UserId": "10186808", "Text": "Changing some of the attributes in DynamoDB is not permitted, for example, changing Partition key, adding a Local Secondary Index, etc. \r\n\r\nWhen such changes occur, Cloudformation will need to replace the resource and to replace, it will try to delete and re-create the resource. During this process, if table already exists, it will fail.\r\n\r\nOnly option is to delete the stack or manually delete the DynamoDB Table and let template create it again.", "ContentLicense": "CC BY-SA 4.0", "filtered-sentences": [{"source": "Text", "text": "Changing some of the attributes in DynamoDB is not permitted, for example, changing Partition key, adding a Local Secondary Index, etc. ", "keywords": ["change"]}, {"source": "Text", "text": "When such changes occur, Cloudformation will need to replace the resource and to replace, it will try to delete and re-create the resource. ", "keywords": ["change"]}]}, {"Id": "241622008", "PostHistoryTypeId": "5", "PostId": "66387833", "RevisionGUID": "81a98c6c-f73e-451d-8638-23a3f025901b", "CreationDate": "2021-02-26T14:48:03.737", "UserId": "10186808", "Comment": "deleted 12 characters in body", "Text": "Changing some of the attributes in DynamoDB is not permitted, for example, changing Partition key, adding a Local Secondary Index, etc. \r\n\r\nWhen such changes occur, it will need to replace the resource and to replace, it will try to delete and re-create the resource. During this process, if table already exists, it will fail.\r\n\r\nOnly option is to delete the stack or manually delete the DynamoDB Table and let template create it again.", "ContentLicense": "CC BY-SA 4.0", "filtered-sentences": [{"source": "Text", "text": "Changing some of the attributes in DynamoDB is not permitted, for example, changing Partition key, adding a Local Secondary Index, etc. ", "keywords": ["change"]}, {"source": "Text", "text": "When such changes occur, it will need to replace the resource and to replace, it will try to delete and re-create the resource. ", "keywords": ["change"]}]}, {"Id": "241622201", "PostHistoryTypeId": "5", "PostId": "66387833", "RevisionGUID": "b4413396-0700-4977-a650-e86c8b05ee31", "CreationDate": "2021-02-26T14:51:00.257", "UserId": "10186808", "Comment": "deleted 12 characters in body", "Text": "Changing some of the attributes in DynamoDB is not permitted, for example, changing Partition key, adding a Local Secondary Index, etc. \r\n\r\nWhen such changes occur, it will need to replace the resource and to replace, it will try to delete and re-create the resource. During this process, if table already exists, it will fail.\r\n\r\nOnly option is to delete the stack or manually delete the DynamoDB Table and let template create it again.\r\n\r\n[Documentation][1] says it will force new resource\r\n\r\n\r\n> hash_key - (Required, Forces new resource) The attribute to use as the\r\n> hash (partition) key.\r\n\r\n  [1]: https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/dynamodb_table", "ContentLicense": "CC BY-SA 4.0", "filtered-sentences": [{"source": "Text", "text": "Changing some of the attributes in DynamoDB is not permitted, for example, changing Partition key, adding a Local Secondary Index, etc. ", "keywords": ["change"]}, {"source": "Text", "text": "When such changes occur, it will need to replace the resource and to replace, it will try to delete and re-create the resource. ", "keywords": ["change"]}]}, {"Id": "241622714", "PostHistoryTypeId": "5", "PostId": "66387833", "RevisionGUID": "56205dc6-48fe-4f90-8e31-c07b94ed8393", "CreationDate": "2021-02-26T14:58:25.823", "UserId": "10186808", "Comment": "added 23 characters in body", "Text": "Changing some of the attributes in DynamoDB is not permitted, for example, changing Partition key, adding a Local Secondary Index, etc. \r\n\r\nWhen such changes occur, it will need to replace the resource and to replace, it will try to delete and re-create the resource. During this process, if table already exists, it will fail.\r\n\r\nOnly option is to delete the stack or manually delete the DynamoDB Table and let template create it again. Or Renaming the table.\r\n\r\n[Documentation][1] says it will force new resource\r\n\r\n\r\n> hash_key - (Required, Forces new resource) The attribute to use as the\r\n> hash (partition) key.\r\n\r\n  [1]: https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/dynamodb_table", "ContentLicense": "CC BY-SA 4.0", "filtered-sentences": [{"source": "Text", "text": "Changing some of the attributes in DynamoDB is not permitted, for example, changing Partition key, adding a Local Secondary Index, etc. ", "keywords": ["change"]}, {"source": "Text", "text": "When such changes occur, it will need to replace the resource and to replace, it will try to delete and re-create the resource. ", "keywords": ["change"]}]}], "filtered-sentences": [{"source": "Body", "text": "Changing some of the attributes in DynamoDB is not permitted, for example, changing Partition key, adding a Local Secondary Index, etc. ", "keywords": ["change"]}, {"source": "Body", "text": "When such changes occur, it will need to replace the resource and to replace, it will try to delete and re-create the resource. ", "keywords": ["change"]}]}], "contains-topic": true, "filtered-sentences": [{"source": "Title", "text": "Changing hash_key with Terraform causes ", "keywords": ["change"]}, {"source": "Body", "text": "I have a DynamoDB table created with this Terraform: The table was successfully populated (with 4 records, as noted in this post) but in order to solve the problem (in that post) I have added a field PK and set that as the hash_key field, with this: This has caused the following error, when running terraform apply: What do I need to do in the .tf to get the change accepted?", "keywords": ["change"]}]}